Atomic/syscontainers.py

```diff
@@ -2131,84 +2127,56 @@ def get_layers_from_manifest(manifest):
-                    key = "%s-%s" % (s.st_dev, s.st_ino)
-                    if key not in seen:
-                        seen[key] = key
-                        size += s.st_size
-            return GLib.Variant('s', str(size))
-
-        repo.prepare_transaction()
-        for layer, tar in layers.items():
-            mtree = OSTree.MutableTree()
-            def filter_func(*args):
-                info = args[2]
-
-                if info.get_file_type() == Gio.FileType.SPECIAL:
-                    return OSTree.RepoCommitFilterResult.SKIP
-
-                if info.get_file_type() == Gio.FileType.DIRECTORY:
-                    info.set_attribute_uint32("unix::mode", info.get_attribute_uint32("unix::mode") | stat.S_IWUSR)
-                return OSTree.RepoCommitFilterResult.ALLOW
-
-            modifier = OSTree.RepoCommitModifier.new(0, filter_func, None)
-
-            checkout = self._get_system_checkout_path()
-            destdir = checkout if os.path.exists(checkout) else None
+    def _pull_docker_image(self, repo, image):
+        with tempfile.NamedTemporaryFile(mode="w") as temptar:
+            util.check_call(["docker", "save", "-o", temptar.name, image])
+            return self._pull_docker_tar(repo, temptar.name, image)
 
-            try:
-                temp_dir = tempfile.mkdtemp(prefix=".", dir=destdir)
-                # NOTE: tarfile has an issue with utf8. This works around the problem
-                # by using the systems tar command.
-                # Ref: https://bugzilla.redhat.com/show_bug.cgi?id=1194473
-                subprocess.check_call(['tar', '-xf', tar, '-C', temp_dir])
-                if self.user:
-                    SystemContainers._correct_dir_permissions_for_user(temp_dir)
-
-                repo.write_directory_to_mtree(Gio.File.new_for_path(temp_dir), mtree, modifier)
-                root = repo.write_mtree(mtree)[1]
-
-                metav = GLib.Variant("a{sv}", {'docker.layer': GLib.Variant('s', layer),
-                                               'docker.size': get_directory_size(temp_dir)})
-                csum = repo.write_commit(None, "", None, metav, root)[1]
-            finally:
-                shutil.rmtree(temp_dir, ignore_errors=True)
+    def _pull_docker_tar(self, repo, tarpath, image):
+        image_name = SystemContainers._get_image_name_from_dockertar(tarpath) or image
+        imagebranch = "%s%s" % (OSTREE_OCIIMAGE_PREFIX, SystemContainers._encode_to_ostree_ref(image_name))
+        image_rev = repo.resolve_rev(imagebranch, True)
 
-            repo.transaction_set_ref(None, "%s%s" % (OSTREE_OCIIMAGE_PREFIX, layer), csum)
+        # Before copying into ostree, we want to first verify if image with same name exists or not
+        # to avoid accidental image overrides
+        if not image_rev:
+            raise ValueError("Internal error for image: %s.  Please pull the image again" % imagebranch)
+        if image_rev[1]:
```

Comment:
This line checks if the image rev exists in ostree backend. But it also causes the unit test failure in `test_pull.py`. Not sure why that happened yet...